File: services/member/src/telemetry.js
/**
 * OpenTelemetry 集成 - 会员服务
 * Traces → OTLP → Tempo
 * Metrics → Prometheus
 * Logs → 结构化 JSON（由 Loki 采集）
 */
'use strict';

function setupTelemetry() {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  const serviceName = process.env.OTEL_SERVICE_NAME || 'member-svc';

  if (!endpoint) {
    process.stderr.write(JSON.stringify({
      timestamp: new Date().toISOString(),
      level: 'WARN',
      service: serviceName,
      message: 'OTEL_EXPORTER_OTLP_ENDPOINT not set, telemetry disabled',
    }) + '\n');
    return;
  }

  try {
    const { NodeSDK } = require('@opentelemetry/sdk-node');
    const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
    const { Resource } = require('@opentelemetry/resources');
    const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');
    const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');

    const sdk = new NodeSDK({
      resource: new Resource({
        [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
        [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
        'deployment.environment': process.env.APP_ENV || 'dev',
      }),
      traceExporter: new OTLPTraceExporter({ url: endpoint }),
      instrumentations: [getNodeAutoInstrumentations()],
    });

    sdk.start();

    process.on('SIGTERM', () => sdk.shutdown());

    process.stdout.write(JSON.stringify({
      timestamp: new Date().toISOString(),
      level: 'INFO',
      service: serviceName,
      message: 'OpenTelemetry initialized',
      endpoint,
    }) + '\n');
  } catch (e) {
    process.stderr.write(JSON.stringify({
      timestamp: new Date().toISOString(),
      level: 'WARN',
      service: serviceName,
      message: `OpenTelemetry setup failed: ${e.message}. Continuing without telemetry.`,
    }) + '\n');
  }
}

module.exports = { setupTelemetry };

File: services/reservation/src/telemetry.py
"""OpenTelemetry 集成配置 - 预约服务"""
import os
import logging

logger = logging.getLogger(__name__)


def setup_telemetry() -> None:
    """
    初始化 OpenTelemetry 三件套：
    - Traces → OTLP → Tempo
    - Metrics → Prometheus Exporter
    - Logs → 结构化 JSON（通过 Loki 采集）
    """
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "")
    service_name = os.getenv("OTEL_SERVICE_NAME", "reservation-svc")

    if not endpoint:
        logger.warning("OTEL_EXPORTER_OTLP_ENDPOINT not set, telemetry disabled")
        return

    try:
        from opentelemetry import trace
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
        from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor

        # 资源标签
        resource = Resource.create({
            "service.name": service_name,
            "service.version": "1.0.0",
            "deployment.environment": os.getenv("APP_ENV", "dev"),
        })

        # Tracer Provider
        tracer_provider = TracerProvider(resource=resource)
        otlp_exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
        tracer_provider.add_span_processor(BatchSpanProcessor(otlp_exporter))
        trace.set_tracer_provider(tracer_provider)

        # 自动埋点
        FastAPIInstrumentor().instrument()
        SQLAlchemyInstrumentor().instrument()

        logger.info("OpenTelemetry initialized", extra={"endpoint": endpoint, "service": service_name})

    except ImportError as e:
        logger.warning(f"OpenTelemetry packages not installed: {e}. Continuing without telemetry.")
